12518 move seo to bundle

mattamon opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `10.6`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.6` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves #12518 

## Additional info  
Moving seo functionality into SeoBundle
Able to install and uninstall via Installer. Permissions are in the Bundle
User Keybindings will be added after glossary bundle is merged
Moved translations into the bundle

mattamon committed:
Extract seo from admin

mattamon committed:
Readd seo menu via bundle

mattamon committed:
Adding new routes, installer and migration and templates

mattamon committed:
Adding translations
Removing keybinding helpers (will be reintroduced with new event)

mattamon committed:
Add SeoBundle to ci/files/config/bundles.php

mattamon committed:
Adding feedback changes
Adding warning to enable bundle manually
Removing unnecessary interfaces
Changing route names

mattamon committed:
Adding feedback changes
Reverting routes
Adding bundle name to migration warning

kingjia90 commented on bundles/AdminBundle/src/Controller/Admin/SettingsController.php:
```diff
@@ -1553,49 +1553,6 @@ public function videoThumbnailUpdateAction(Request $request): JsonResponse
-        }
-
-        foreach ($values as $siteId => $robotsContent) {
-            SettingsStore::set('robots.txt-' . $siteId, $robotsContent, 'string', 'robots.txt');
```
the following import can be removed
https://github.com/pimcore/pimcore/blob/ecd4a3f54bfa521916c5603f649d5131e34a54d0/bundles/AdminBundle/src/Controller/Admin/SettingsController.php#L38

kingjia90 commented on bundles/SeoBundle/public/js/seopanel.js:
```diff
@@ -21,6 +21,7 @@ pimcore.document.seopanel = Class.create({
 
     activate: function () {
         var tabPanel = Ext.getCmp("pimcore_panel_tabs");
+        console.log(tabPanel);
```
```suggestion

```

kingjia90 commented on bundles/SeoBundle/src/Installer.php:
```diff
@@ -0,0 +1,63 @@
+        $db = \Pimcore\Db::get();
+
+        foreach (self::USER_PERMISSIONS as $permission) {
+            $db->insert('users_permission_definitions', [
```
in my local instance, this would break because these permissions definitions are already there
Can we change it to something that check if already exists? and/or a migration to remove the existing permissions

mattamon replied:
I moved the migration to the CoreBundle. It should by default get marked as installed, so the permission should not be installed again. If you do a fresh install, this problem should not occur, because I removed the permissions in the InstallerBundle. So basically when you upgrade, the bundle should be marked as installed by default. 

Wdyt?

kingjia90 replied:
sounds good to me, at least for a seamless upgrade from 10.x, ~but not sure about from a skeleton point of view, if anything "optional" should be deactiveted by default. Let's ask in the daily~ confrimed that by default in fresh install it should be deactived

kingjia90 commented on bundles/SeoBundle/public/js/startup.js:
```diff
@@ -0,0 +1,96 @@
+            if (seoMenu.length > 0) {
+                const toolbar = pimcore.globalmanager.get('layout_toolbar');
+
+                toolbar.marketingMenu.add({
```
For a non-admin user with only SEO bundle permissions, the `marketingMenu` doesn't exists (due the IF in the following snippet), and thus won't work/appear.
 
https://github.com/pimcore/pimcore/blob/ecd4a3f54bfa521916c5603f649d5131e34a54d0/bundles/AdminBundle/public/js/pimcore/layout/toolbar.js#L580-L581

mattamon replied:
You are right, maybe we should discuss, what should happen here then. 
There are some changes with the PimcoreGlossaryBundle but they still would not solve this problem.
https://github.com/pimcore/pimcore/pull/13811/files#diff-0bd635bc68ae1ed97161f850140cef5707f9be71d5dd982c1f978c6aab2255c7R604

Maybe we should add those menus even though they are empty and check before creating the menu again.
https://github.com/pimcore/pimcore/pull/13811/files#diff-0bd635bc68ae1ed97161f850140cef5707f9be71d5dd982c1f978c6aab2255c7R1156 -> with this event you can still add items. Maybe we have a check after that if some menu items would be empty right? 

Wdyt?

mattamon replied:
Like we discussed in the daily:
https://github.com/pimcore/pimcore/pull/13811/files#diff-0bd635bc68ae1ed97161f850140cef5707f9be71d5dd982c1f978c6aab2255c7R482 -> i removed the if condition.
Everyone who does the refactoring should do this, so we do it one by one.

Unfortunately it is not possible to simply hide the menu item so I do not create it if items are empty.
https://github.com/pimcore/pimcore/pull/13811/files#diff-0bd635bc68ae1ed97161f850140cef5707f9be71d5dd982c1f978c6aab2255c7R1164

kingjia90 replied:
what about moving all the code in a separate method 
eg. 
```
createMarketingMenu: function(items){
   this.marketingMenu = new Ext.menu.Menu({
   ...
}
```
and from the bundle, check if `typeof toolbar.marketingMenu == 'undefined'` , then `toolbar.createMarketingMenu()` first, and finally `toolbar.marketingMenu.add();`

mattamon replied:
Like I mentioned, I refactored everything in the GlossaryBundle, so maybe we should wait for those changes to be merged. :) This add method won't work anymore because we create every menu at the very end, to avoid sorting the menu after it is already rendered.

kingjia90 reviewed (changes requested):
Overall looks and works fine, just found some minor issues

mattamon committed:
Renaming SeoBundle to PimcoreSeoBundle

mattamon committed:
Moving translations back into core

mattamon committed:
Merge branch '11.x' into 12518-move-seo-to-bundle

mattamon committed:
Merge branch '11.x' into 12518-move-seo-to-bundle

# Conflicts:
#	.github/ci/files/config/bundles.php
#	composer.json

mattamon committed:
Added upgrade notes
Added menu items
Added key bindings

mattamon committed:
Changing js class namespace, routing prefix and routing names for consistency
Moving http_error_log table to bundle and installer
Adding Exceptionlistener with higher priority

mattamon committed:
Rename pimcoreReady function to preMenuBuild

mattamon committed:
Remove unecessary import

mattamon committed:
Refactor JS namespaces and extracting translations to admin.en.json

mattamon committed:
Merge branch '11.x' into 12518-move-seo-to-bundle

brusch commented on bundles/SeoBundle/src/Controller/Document/DocumentController.php:
```diff
@@ -0,0 +1,162 @@
+{
+    use DocumentTreeConfigTrait;
+
+    final const DOCUMENT_ROOT_ID = 1;
```
```suggestion
    final private const DOCUMENT_ROOT_ID = 1;
```

mattamon replied:
I made it private only, final private is not possible, since private is not visible to others.

brusch commented:
@mattamon there are some new conflicts that need to be resolved 😊

mattamon committed:
Merge branch '11.x' into 12518-move-seo-to-bundle

# Conflicts:
#	bundles/SeoBundle/public/js/httpErrorLog.js
#	bundles/SeoBundle/public/js/robotstxt.js
#	bundles/SeoBundle/public/js/seopanel.js

mattamon committed:
Javascript: tag private classes

mattamon committed:
Making constant DOCUMENT_ROOT_ID privat

kingjia90 commented on bundles/SeoBundle/public/js/startup.js:
```diff
@@ -0,0 +1,130 @@
+            pimcore.globalmanager.get("bundle_seo_seo_seopanel").activate();
+        }
+        catch (e) {
+            pimcore.globalmanager.add("bundle_seo_seo_seopanel", new pimcore.bundle.seo.seopanel());
```
Looks like there's an extra `seo` 
```suggestion
            pimcore.globalmanager.add("bundle_seo_seopanel", new pimcore.bundle.seo.seopanel());
```

kingjia90 commented on bundles/SeoBundle/src/Controller/SettingsController.php:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        foreach ($values as $siteId => $robotsContent) {
+            SettingsStore::set('robots.txt-' . $siteId, $robotsContent, 'string', 'robots.txt');
```
Not sure if something changed in this PR, but i cannot see the `robots.txt`
On demo.pimcore.fun, in the settingstore table, id column is `robots.txt-default`, while on this is `robots.txt-0`. 
Could you please also have a look.

mattamon replied:
I just took a deeper look and it this was part of a type fix PR
https://github.com/pimcore/pimcore/commit/92da8b98e56ef80ddf4e5cefa5fdb78b7af999f9#diff-d8a9d98ddf01d40976e2f54e4ed4c36bdd81218f67107385e81f3947911b7828R1044
There it changed from "default" to 0. 
Maybe we should bring it up?

kingjia90 reviewed (changes requested):
LGTM, just found these 2 things, beside the keybidings which will be fixed in https://github.com/pimcore/pimcore/issues/13994

mattamon committed:
Remove additonal seo string
